### ISSUE TITLE
[Question] Forking of CoreCLR process

### ISSUE THREAD

**Issue by gbalykov** [dotnet/coreclr#13936]
Hello

Does CoreCLR supports forking of the process with initialized CoreCLR and further independent operation of the CoreCLR instances?

**Comment by gbalykov** [comment]
@jkotas @janvorli

Please, take a look

**Comment by jkotas** [comment]
I doubt that it will work. It is certainly not officially supported.

**Comment by gbalykov** [comment]
@jkotas 

What do you think about extending CoreCLR hosting API with alternative initialization: `coreclr_initialize_prefork` and `coreclr_initialize_postfork`, first of which will not launch any threads, and the second one will?

**Comment by jkotas** [comment]
What it would be used for?

**Comment by gbalykov** [comment]
@jkotas

We would like to implement the mechanism similar to Zygote in Dalvik.

CoreCLR will be initialized with `coreclr_initialize_prefork`, which will not launch any threads. Then fork() will be called, and both parent and child processes will call `coreclr_initialize_postfork`, which will finish initialization of CoreCLR for each process. After that each process will be able to call `coreclr_execute_assembly` with its own set of parameters. 

`coreclr_initialize_prefork` will be similar to `initZygote` in Dalvik and `coreclr_initialize_postfork` will be similar to `dvmInitAfterZygote` in Dalvik.

**Comment by jkotas** [comment]
Extra threads get created pretty early. What is that you are trying to save?

Also, it is not just about threads. There are other ways to get incompatible with fork, e.g. cache or take dependency on pid.

**Comment by gbalykov** [comment]
@jkotas 

We want to preload assemblies before fork, so that they are shared after forking. 

What caches do you mean?

We could try to remove the dependecy on saved process ID, what do you think about this?

**Comment by jkotas** [comment]
The first thread gets created today before any assemblies are loaded (even before CoreLib).

Is preloading of CoreLib the only thing you are going after?

**Comment by gbalykov** [comment]
@jkotas 

We would like to load several dlls, which occupy most of the memory on particular set of applications. If it is possible to extend hosting API, we would also like to add loading of other assemblies before any threads are created.

**Comment by jkotas** [comment]
Yes, that would make sense. I do not think that the API you are proposing helps you achieve that.

**Comment by gbalykov** [comment]
Yes, the two API functions above are not enough. The full list of extensions could look like this:

- `coreclr_initialize_prefork`
- `coreclr_preload_assembly`
- `coreclr_initialize_postfork` (fork happens before this phase, threads are created upon reaching this phase)

**Comment by jkotas** [comment]
What would `coreclr_preload_assembly` do? Just map the .dll into memory, or more?

**Comment by gbalykov** [comment]
`coreclr_preload_assembly` will map dlls, register them, so that loader could find them, and apply relocations

**Comment by jkotas** [comment]
Can we have just `core_preload_assembly` ? You would call this before calling regular `coreclr_initialize`. The implementation would map the file, apply relocations and store the result in a linked list. This will run relatively small amount of code that one can reason about and make fork friendly. The regular load would then consult the linked list.

**Comment by gbalykov** [comment]
Preinitialization of CoreCLR before forking could give additional benefits because of sharing of execution engine and type information of non-precompiled assemblies.

We agree that starting with just `core_preload_assembly` is very good idea, because it should save significant amount of memory and it is much simpler to implement.

We will try to implement this and open pull request in next two weeks.

Thank you for your responses!

**Comment by alpencolt** [comment]
Android Runtime has several threads like CoreCLR but before forking zygote it stop all except main and after that restart them. Is it difficult to do the same in CoreCLR?

@jkotas how many threads work when we execute simple Hello world application?

**Comment by jkotas** [comment]
There is the main thread, finalizer thread, debugger thread, background JITing thread, PAL eventing  thread, ... .
